gdal/ogr/ogrpolyhedralsurface.cpp

```diff
@@ -650,3 +703,184 @@ OGRMultiPolygon* OGRPolyhedralSurface::CastToMultiPolygon()
     OGRMultiPolygon *poMultiPolygon = new OGRMultiPolygon(this->oMP);
     return poMultiPolygon;
 }
+
+/************************************************************************/
+/*                            addGeometry()                             */
+/*      Add a new geometry to a collection.  Subclasses should          */
+/*      override this to verify the type of the new geometry, and       */
+/*      then call this method to actually add it.                       */
+/************************************************************************/
+
+OGRErr OGRPolyhedralSurface::addGeometry (const OGRGeometry *poNewGeom)
+{
+    if (!EQUAL(poNewGeom->getGeometryName(),"POLYGON"))
+        return OGRERR_UNSUPPORTED_GEOMETRY_TYPE;
+
+    OGRGeometry *poClone = poNewGeom->clone();
+    OGRErr      eErr;
+
+    if (poClone == NULL)
+        return OGRERR_FAILURE;
+    eErr = oMP.addGeometryDirectly(poClone);
+    if( eErr != OGRERR_NONE )
+        delete poClone;
+
+    return eErr;
+}
+
+/************************************************************************/
+/*                           getNumPolygons()                           */
+/************************************************************************/
+
+int OGRPolyhedralSurface::getNumPolygons()
+{
+    return oMP.nGeomCount;
+}
+
+/************************************************************************/
+/*                             getPolygon()                             */
+/************************************************************************/
+
+OGRGeometry* OGRPolyhedralSurface::getPolygon(int i)
+{
+    return (OGRGeometry *)oMP.papoGeoms[i];
+}
+
+/************************************************************************/
+/*                               IsEmpty()                              */
+/************************************************************************/
+
+OGRBoolean  OGRPolyhedralSurface::IsEmpty() const
+{
+    return oMP.IsEmpty();
+}
+
+/************************************************************************/
+/*                                 set3D()                              */
+/************************************************************************/
+
+void OGRPolyhedralSurface::set3D (OGRBoolean bIs3D)
+{
+    for( int iGeom = 0; iGeom < oMP.nGeomCount; iGeom++ )
```

Comment:
why not calling oMP.set3D( bIs3D ) ? Similar remark for setMeasured(), setCoordinateDimension(). swapXY(), etc... Avoid as much as possible to use the internal details of OGRMultiPolygon (the fact it has a papoGeoms member for example)  when you can